### PR TITLE
Fix scheduled benchmarks to test both last release and main

### PR DESCRIPTION
**Pull request by Czaki**
The scheduled benchmarks are broken. In #6278 I fixed running benchmarks against the main branch. But I missed that benchmarks must also run against the last release (currently 0.4.19). 

This PR adds checking benchmarks against the latest release in pull_request tests. 

For any API that was changed, this PR adds a conditional `if` branch and uses the old api for 0.4.19 and the new one for main. This should simplify after removing obsolete code in 0.5.0 release. 

This PR enables Async benchmarks, but only on the main branch. 

I also renamed the previously misspelled `Skiper` class to `Skip` and added constructor arguments to provide predicate functions about *when* to skip:

* `if_in_pr`: used to skip benchmarks if the `PR` environment variable is defined.
* `if_in_ci`: used to skip benchmarks if the `CI` environment variable is defined. (Because CI has reduced RAM compared to most machines.)
* `always`: used to skip incomplete benchmarks, or to skip last-release benchmarks on new functions.

### CODE REVIEW

File: napari/benchmarks/benchmark_shapes_layer.py [hunk 1]
```diff
@@ -2,18 +2,24 @@
...
 from napari.layers import Shapes
-from napari.utils._test_utils import read_only_mouse_event
 from napari.utils.interactions import (
     mouse_move_callbacks,
     mouse_press_callbacks,
     mouse_release_callbacks,
 )
 
+# from napari.utils._test_utils import read_only_mouse_event
+# could be uncommented after release 0.4.19
```
Comment (jni) [inline napari/benchmarks/benchmark_shapes_layer.py]: ?

**Comment by jni** [comment]
@Czaki can you summarise what the fixes were? Since the Skipper was renamed to fix spelling (thanks), and some of the conditions were changed, it's hard to tell what was broken in 0.4.18 and how this fixes things. Also, is this run against 0.4.19 now rather than 0.4.18? Can you update the description for this?

File: .github/workflows/benchmarks_report.yml [hunk 2]
```diff
@@ -18,6 +18,7 @@ on:
 jobs:
   download:
     runs-on: ubuntu-latest
+    if: ${{ github.event.workflow_run }}
```
Comment (jni) [inline .github/workflows/benchmarks_report.yml]: What does this line do in this context? It might be worth adding a comment to it...

File: .github/workflows/benchmarks.yml [hunk 3]
```diff
@@ -62,6 +62,10 @@ jobs:
       - uses: actions/checkout@v4
         with:
           fetch-depth: 0
+      - name: add upstream repository
+        run:  |
+          git remote add upstream https://github.com/napari/napari.git
+          git fetch --all
```
Comment (jni) [inline .github/workflows/benchmarks.yml]: Why is this needed? How does it change the behaviour of the code below? Is it because we were missing a branch?

File: .github/workflows/test_pull_requests.yml [hunk 4]
```diff
@@ -187,25 +187,51 @@ jobs:
...
           PR: 1
+
+      - name: Run benchmarks latest release
+        # here we check if the benchmark on the latest release is not broken
+        uses: aganders3/headless-gui@v1
+        with:
+          run: |
+            asv run --show-stderr --quick  --attribute timeout=300  ${{ fromJSON(steps.latest_release.outputs.data).target_commitish }}^!
+        env:
+          PR: 1
```
Comment (jni) [inline .github/workflows/test_pull_requests.yml]: 1?

I'd love some guidance here, perhaps in a comment, about what to do if the benchmarks break in your PR because of an API change...

File: napari/benchmarks/benchmark_labels_layer.py [hunk 5]
```diff
@@ -144,19 +144,22 @@ class Labels3DSuite:
...
     param_names = ['n', 'dtype']
     params = ([2**i for i in range(4, 11)], [np.uint8, np.uint32])
-    if 'PR' in os.environ:
-        skip_params = [(2**i,) for i in range(6, 11)]
 
-    def setup(self, n, dtype):
-        if 'CI' in os.environ and n > 512:
-            raise NotImplementedError('Skip on CI (not enough memory)')
+    skip_params = Skipper(
+        func_pr=lambda x: x[0] > 2**6, func_ci=lambda x: x[0] > 2**9
```
Comment (jni) [inline napari/benchmarks/benchmark_labels_layer.py]: what is "x"? It would be better if the variable name could be more descriptive... Otherwise, some comments here would be useful, e.g. "skip if image total size is greater than `2**6`" (although I think this is side length, not size?)

File: napari/benchmarks/benchmark_labels_layer.py [hunk 6]
```diff
@@ -144,19 +144,22 @@ class Labels3DSuite:
...
         np.random.seed(0)
         self.data = labeled_particles(
             (n, n, n), dtype=dtype, n=int(np.log2(n) ** 2), seed=1
         )
         self.layer = Labels(self.data)
-        self.layer._slice_dims(Dims(ndim=3, ndisplay=3))
+        if NAPARI_0_4_19:
+            self.layer._slice_dims((0, 0, 0), ndisplay=3)
+        else:
+            self.layer._slice_dims(Dims(ndim=3, ndisplay=3))
```
Comment (jni) [inline napari/benchmarks/benchmark_labels_layer.py:164]: This kind of code makes me wonder whether we should even bother testing the last release (rather than only main, and therefore PRs on main). It seems like it will be a significant burden to maintain while we are still evolving the API...

File: napari/benchmarks/benchmark_points_layer.py [hunk 7]
```diff
@@ -136,8 +141,7 @@ class PointsToMaskSuite:
         [5, 10],
     ]
 
-    if 'PR' in os.environ:
-        skip_params = Skiper(lambda x: x[0] > 256 or x[1][0] > 512)
+    skip_params = Skipper(func_pr=lambda x: x[0] > 256 or x[1][0] > 512)
```
Comment (jni) [inline napari/benchmarks/benchmark_points_layer.py]: I'm having trouble parsing `func_pr=` and `func_ci=`. Maybe we can rename the class to `Skip` and the parameters `in_ci_if=` and `in_pr_if=`? That would read something like:

```
skip_params = Skip(in_pr_if=lambda shp: shp[0] > 256, in_ci_if=lambda shp: shp[0] > 512)
```

That starts to be a bit more explanatory I think. What do you think?

**Review by jni — COMMENTED** [review]
Thanks!

I've left a few comments to clarify this. I think when they are addressed, this could go in. Overall, though, I'm not super sure this is worth it... The benchmark code is now much more complicated, and this will require constant churn whenever we change APIs. Maybe it is worth just having main work and ignoring previous releases? (And I would say this is *especially* true for 0.4.19...)

File: .github/workflows/test_pull_requests.yml (hunk 4 above)
Comment (Czaki) [reply .github/workflows/test_pull_requests.yml]: `${{ fromJSON(steps.latest_release.outputs.data).target_commitish }}` is getting output of `latest_release` steep and extract from it hash of release commit.

`1` is same as in `Run benchmarks PR` steep. And is used to indicate to omit long computing case (and reduce time from 10h to 6 minutes).

File: napari/benchmarks/benchmark_labels_layer.py (hunk 6 above)
Comment (Czaki) [reply napari/benchmarks/benchmark_labels_layer.py:164]: Yes, we need to trace performance degradation since previous release.

File: .github/workflows/benchmarks.yml (hunk 3 above)
Comment (Czaki) [reply .github/workflows/benchmarks.yml]: Will remove it, as it is connected with try to fix reporting. I will make followup PR for reporting

File: .github/workflows/benchmarks_report.yml (hunk 2 above)
Comment (Czaki) [reply .github/workflows/benchmarks_report.yml]: Will remove it, as it is connected with try to fix reporting. I will make followup PR for reporting

**Comment by Czaki** [comment]
> Maybe it is worth just having main work and ignoring previous releases?

Then why even have benchmarks if we cannot see notice degradation of napari performance? 

> And I would say this is _especially_ true for 0.4.19...

Making it compatible with any older release will be a nightmare. 

The proper approach will be to not use private API in benchmarks. However, part of the changes comes from breaking of public API without any transition period. 

For example https://github.com/napari/napari/pull/6496/files#diff-d4011f25abcada1d5bc826008c525db1813684120c2ec28c7ac5ee3ddd33d5efR79-R92  (rember that `QtViewer` is public class). 

So I may fix this cases in code by adding deprecated methods. That we could use in benchamrks.

File: napari/benchmarks/benchmark_labels_layer.py (hunk 5 above)
Comment (Czaki) [reply napari/benchmarks/benchmark_labels_layer.py]: I have changed Skipper code to pass unpacked tuple and unpack it and use more descriptive names.

File: napari/benchmarks/benchmark_points_layer.py (hunk 7 above)
Comment (Czaki) [reply napari/benchmarks/benchmark_points_layer.py]: I am not convinced that `in_pr_if` is more readable than `func_pr`, especially in `func_pr` you do not check the middle of the word to find which case is covered here` 

Maybe `if_in_pr`, `if_on_ci` and `always`?

Comment (jni) [reply napari/benchmarks/benchmark_points_layer.py]: > Maybe if_in_pr, if_on_ci and always?

Sure.

I'd also rename the class to Skip. For some reason Skipper feels a bit weird — probably because it can mean [many other things](https://www.merriam-webster.com/dictionary/skipper).

File: napari/benchmarks/benchmark_labels_layer.py (hunk 6 above)
Comment (jni) [reply napari/benchmarks/benchmark_labels_layer.py:164]: I suppose. But this can also be achieved by making sure that the benchmarks work in each PR *and* that they are run in any PR that could have performance indication. That is, I would rather spend maintenance effort in our benchmark running *infrastructure* than in playing this API balancing act.

Comment (jni) [reply napari/benchmarks/benchmark_labels_layer.py:164]: (But I'm not super opposed to this — just wondering out loud whether there are better ways, with less maintenance cost, to achieve the goal of monitoring and improving performance.)

Comment (Czaki) [reply napari/benchmarks/benchmark_labels_layer.py:164]: I understand the idea that you want, but it is impossible. The only way to perform such an operation is to have a fixed machine (without any updates and other users). 

When we run benchmarks on CI, the benchmarks cannot be compared between different CI runs.
We sometimes have false negative/positive reports because of instability on a single machine. 

My preferred way is to use stable API in benchmark. As we should do this to provide a transition period for users, then will be win-win.

File: .github/workflows/test_pull_requests.yml (hunk 4 above)
Comment (jni) [reply .github/workflows/test_pull_requests.yml]: > And is used to indicate to omit long computing case (and reduce time from 10h to 6 minutes).

Maybe worth adding a comment to those lines? 

```suggestion
          PR: 1  # prevents asv from running very compute-intensive benchmarks
```

File: napari/benchmarks/benchmark_points_layer.py (hunk 7 above)
Comment (Czaki) [reply napari/benchmarks/benchmark_points_layer.py]: Fixed naming.

File: .github/workflows/test_pull_requests.yml [hunk 8]
```diff
@@ -187,25 +187,51 @@ jobs:
...
+        run: asv machine --yes
+
+      - name: Run benchmarks PR
+        uses: aganders3/headless-gui@v1
         with:
           run: |
-            asv machine --yes
             asv run --show-stderr --quick  --attribute timeout=300 HEAD^!
         env:
           PR: 1
```
Comment (Czaki) [inline .github/workflows/test_pull_requests.yml]:
```suggestion
          PR: 1 # prevents asv from running very compute-intensive benchmarks
```

**Review by jni — APPROVED** [review]
